### PR TITLE
LNX-201 Add a system for adding agents

### PR DESCRIPTION
**Pull request by Krystian030**

### CODE REVIEW

File: src/routers/scene_router.py [hunk 1]
```diff
@@ -5,9 +5,10 @@
 from typing import Dict, Union, List
 
 import opensimplex
-from fastapi import APIRouter
+from fastapi import APIRouter, HTTPException
```
Comment (filip-szweda) [inline src/routers/scene_router.py:8]:
```suggestion
from fastapi import APIRouter, HTTPException

```

File: src/routers/scene_router.py [hunk 2]
```diff
@@ -57,18 +57,51 @@ async def get(tick_number: int, player: str = "") -> Dict[str, Union[int, str, L
 async def add(r: AddObjectDTO, background_tasks: BackgroundTasks):
     logger.debug(f"Starting to add object to scene, starting to deserialize object")
     object = Object.deserialize(r.serialized_object)
+    player = state.scene.get_player(object.owner)
+
+    if player is None:
+        logger.warning(f"Player with id: {object.owner} does not exist")
+        raise HTTPException(status_code=400, detail={"message": f"Player with id: {object.owner} does not exist"})
```
Comment (filip-szweda) [inline src/routers/scene_router.py:64]: Catch this exception explicitly

File: src/routers/scene_router.py [hunk 3]
```diff
@@ -57,18 +57,51 @@ async def get(tick_number: int, player: str = "") -> Dict[str, Union[int, str, L
...
+
+    if player.add_agent_from_object(object) is False:
+        logger.warning(f"Player with id: {object.owner} cannot add agent")
+        cost = player.calculate_slot_cost()
+        raise HTTPException(
```
Comment (filip-szweda) [inline src/routers/scene_router.py:69]: Catch this exception explicitly

File: src/routers/scene_router.py [hunk 4]
```diff
@@ -57,18 +57,51 @@ async def get(tick_number: int, player: str = "") -> Dict[str, Union[int, str, L
 async def add(r: AddObjectDTO, background_tasks: BackgroundTasks):
     logger.debug(f"Starting to add object to scene, starting to deserialize object")
     object = Object.deserialize(r.serialized_object)
+    player = state.scene.get_player(object.owner)
+
+    if player is None:
+        logger.warning(f"Player with id: {object.owner} does not exist")
```
Comment (filip-szweda) [inline src/routers/scene_router.py]:
```suggestion
        logger.error(f"Player with id: {object.owner} does not exist")
```

File: src/routers/scene_router.py [hunk 5]
```diff
@@ -57,18 +57,51 @@ async def get(tick_number: int, player: str = "") -> Dict[str, Union[int, str, L
...
+        raise HTTPException(status_code=400, detail={"message": f"Player with id: {object.owner} does not exist"})
+
+    if player.add_agent_from_object(object) is False:
+        logger.warning(f"Player with id: {object.owner} cannot add agent")
```
Comment (filip-szweda) [inline src/routers/scene_router.py]:
```suggestion
        logger.error(f"Player with id: {object.owner} cannot add agent")
```

File: src/routers/scene_router.py [hunk 6]
```diff
@@ -57,18 +57,51 @@ async def get(tick_number: int, player: str = "") -> Dict[str, Union[int, str, L
...
+    logger.debug(f"Object has been successfully added to pending actions.")
     if object.tick != "":
         logger.debug(f"Object has tick {object.tick}")
         background_tasks.add_task(spawn_process_for_new_agent, object)
     logger.debug(f"Object has been successfully added to scene")
     return {"serialized_object": object.serialize()}
 
 
+@router.post("/buy-slot/{player_id}")
+async def buy_slot(player_id: str):
```
Comment (filip-szweda) [inline src/routers/scene_router.py]:
```suggestion
async def buy_slot(player_id: str) -> None:
```

File: src/routers/scene_router.py [hunk 7]
```diff
@@ -146,20 +179,27 @@ async def get_ranking():
 @router.get("/player-agents/{player_id}")
 async def get_agents_from_player_list(player_id: str):
```
Comment (filip-szweda) [inline src/routers/scene_router.py]:
```suggestion
async def get_agents_from_player_list(player_id: str) -> Dict:
```

**Review by filip-szweda — CHANGES_REQUESTED** [review]

File: src/routers/scene_router.py (hunk 2 above)
Comment (Krystian030) [reply src/routers/scene_router.py:64]: nope

File: src/routers/scene_router.py (hunk 1 above)
Comment (Krystian030) [reply src/routers/scene_router.py:8]: nope

File: src/routers/scene_router.py (hunk 3 above)
Comment (Krystian030) [reply src/routers/scene_router.py:69]: nope

File: src/routers/scene_router.py (hunk 4 above)
Comment (Krystian030) [reply src/routers/scene_router.py]: nope

File: src/routers/scene_router.py (hunk 5 above)
Comment (Krystian030) [reply src/routers/scene_router.py]: nope

File: src/routers/scene_router.py [hunk 8]
```diff
@@ -57,18 +57,51 @@ async def get(tick_number: int, player: str = "") -> Dict[str, Union[int, str, L
...
+        raise HTTPException(status_code=400, detail={"message": f"Player with id: {object.owner} does not exist"})
+
+    if player.add_agent_from_object(object) is False:
+        logger.error(f"Player with id: {object.owner} cannot add agent")
+        cost = player.calculate_slot_cost()
+        raise HTTPException(
+            status_code=400,
+            detail={
+                "message": "Cannot create agent, you should buy more slots",
+                "cost": cost
```
Comment (EasyJB) [inline src/routers/scene_router.py:73]: Cost is too huuuge

**Review by EasyJB — APPROVED** [review]

File: src/routers/scene_router.py (hunk 2 above)
Comment (filip-szweda) [reply src/routers/scene_router.py:64]: https://fastapi.tiangolo.com/tutorial/handling-errors/

File: src/routers/scene_router.py (hunk 3 above)
Comment (filip-szweda) [reply src/routers/scene_router.py:69]: https://fastapi.tiangolo.com/tutorial/handling-errors/

**Review by filip-szweda — APPROVED** [review]

**Pull request merged**